clash-lib/src/Clash/Driver/Manifest.hs

```diff
@@ -0,0 +1,269 @@
+import           System.FilePath (takeDirectory, (</>))
+import           System.Directory (listDirectory, doesFileExist)
+
+import           Clash.Annotations.TopEntity.Extra ()
+import           Clash.Driver.Types
+import           Clash.Primitives.Types
+import           Clash.Core.Var (Id)
+import           Clash.Netlist.Types (TopEntityT, Component(..))
+import qualified Clash.Netlist.Types as Netlist
+import qualified Clash.Netlist.Id as Id
+import           Clash.Primitives.Util (hashCompiledPrimMap)
+import           Clash.Util.Graph (callGraphBindings)
+
+import           Util (OverridingBool(Auto))
+
+-- | Information about the generated HDL between (sub)runs of the compiler
+data Manifest
+  = Manifest
+  { manifestHash :: Int
+    -- ^ Hash of the TopEntity and all its dependencies
+  , successFlags  :: (Int,Int,Bool)
+    -- ^ Compiler flags used to achieve successful compilation:
+    --
+    --   * opt_inlineLimit
+    --   * opt_specLimit
+    --   * opt_floatSupport
+  , portInNames  :: [Text]
+  , portOutNames :: [Text]
+  , componentNames :: [Text]
+    -- ^ Names of all the generated components for the @TopEntity@ (does not
+    -- include the names of the components of the @TestBench@ accompanying
+    -- the @TopEntity@).
+  , topComponent :: Text
+    -- ^ Design entry point. This is usually the component annotated with a
+    -- @TopEntity@ annotation.
+  , fileNames :: [(FilePath, ByteString)]
+    -- ^ Names and hashes of all the generated files for the @TopEntity@. Hashes
+    -- are SHA256.
+  } deriving (Show,Read)
+
+data UnexpectedModification
+  -- | Clash generated file was modified
+  = Modified FilePath
+  -- | Non-clash generated file was added
+  | Added FilePath
+  -- | Clash generated file was removed
+  | Removed FilePath
+
+mkManifest ::
+  -- | Options Clash was run with
+  ClashOpts ->
+  -- | Component of top entity
+  Component ->
+  -- | All other entities
+  [Component] ->
+  -- | Files and  their hashes
+  [(FilePath, ByteString)] ->
+  -- | Hash returned by 'readFreshManifest'
+  Int ->
+  -- | New manifest
+  Manifest
+mkManifest ClashOpts{..} Component{..} components files topHash = Manifest
+  { manifestHash = topHash
+  , portInNames = map Id.toText topInNames
+  , portOutNames = map Id.toText topOutNames
+  , componentNames = map Id.toText compNames
+  , topComponent = Id.toText componentName
+  , fileNames = files
+  , successFlags = (opt_inlineLimit, opt_specLimit, opt_floatSupport)
+  }
+ where
+  topInNames = map fst inputs
+  topOutNames = map (fst . (\(_,x,_) -> x)) outputs
+  compNames = map Netlist.componentName components
+
+-- | Pretty print an unexpected modification as a list item.
+pprintUnexpectedModification :: UnexpectedModification -> String
+pprintUnexpectedModification = \case
+  Modified p -> "Unexpected modification in " <> show p
+  Added p -> "Unexpected extra file " <> show p
+  Removed p -> "Unexpected removed file " <> show p
```

Comment:
I'm inclined to suggested these `show p` be changed to just `p`, normally when you see paths printed on the command line they don't have the surrounding double quotes